### ISSUE TITLE
Calling mock methods w/o default action set and w/o default value set shall fail the test

### ISSUE THREAD

**Issue by PiotrNycz** [google/googletest#4063]
There is one exception that might be thrown by gmock and it can be caught by code-under-test, but it does not mark impacted test as failed.

This is the exception with the following "what()": "The mock function has no default action set, and its return type has no default value set."
As described here: https://github.com/google/googletest/blob/main/docs/gmock_faq.md#i-cant-figure-out-why-gmock-thinks-my-expectations-are-not-satisfied-what-should-i-do

This behavior might cause that wrongly written test might be treated as perfectly valid.
See example: https://godbolt.org/z/fjMTbTsxY

```
#include <gtest/gtest.h>
#include <gmock/gmock.h>

struct A
{
    A(bool a) : a(a) {}
    bool a{};
};

struct Foo {
    MOCK_METHOD(A, foo, ());
};
struct Bar {
    MOCK_METHOD(void, bar, ());
};

void tryFoo(Foo& foo, Bar& bar)
{
    try {
        if (not foo.foo().a) // here exception "The mock function has no default action set, and its return type has no default value set." is thrown
            bar.bar();
    } catch (std::exception const& ex) {
        std::cerr << ex.what(); // and it is caught here
    }
}

TEST(tryFooTest, shallNotCallBarOnDisabledA) {
    ::testing::NiceMock<Foo> foo;
    ::testing::StrictMock<Bar> bar; 
    tryFoo(foo, bar);
}

int main(int argc, char*argv[]) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

```

So - developer thinks the tests correctly test that bar.bar() is not called because of A::a being true. But what actually happen is internal gmock exception causes this test "passes" - but it should not.

The output is as follows:

> 
> Function call: foo
>     The mock function has no default action set, and its return type has no default value set.
> [==========] Running 1 test from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 1 test from tryFooTest
> [ RUN      ] tryFooTest.shallNotCallBarOnDisabledA
> [       OK ] tryFooTest.shallNotCallBarOnDisabledA (0 ms)
> [----------] 1 test from tryFooTest (0 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (0 ms total)
> [  PASSED  ] 1 test.

What I suggest is to change this behavior -- I mean - to mark any test that throws this internal exception as failed.
To enable smooth upgrade to fixed version of gmock -- some kind of global bool variables might be introduced which might "restore" the old, before-fix, wrong behavior.

**Comment by asoffer** [comment]
Thanks for raising the issue. I think part of the problem is in a world with exceptions, catching an exception has to be considered valid and intended ways to handle an exceptional situation. For that reason it doesn't make sense to treat this as an error: By attempting to catch the issue, you've indicated that you know how to handle the issue.

**Comment by PiotrNycz** [comment]
@asoffer I think you missed the point here. The point is:

- gmock detects some wrong situation - call of mocked method with return value T - and this T has no default constructor nor its default values is set is some other ways allowed by gmock
- gmock throws exception - indicating something is wrong
- this exception IS NOT what code-under-text is expected nor prepared for
- my interpretation is entire situation is that gmock surely thinks the test fails
- and it is not fails - so gmock behavior is WRONG and needs to be fixed

**Comment by PiotrNycz** [comment]
@asoffer  Maybe I have not said it openly in my last comment, I strongly advise to consider to **RE-OPEN** this issue. I believe it should be fixed. If not fixed, people will continue to write wrong(fooling) tests with using gmock.

**Comment by boimart1** [comment]
FYI, this issue was incidentally fixed in 9d21db9e0a60a1ea61ec19331c9bc0dd33e907b1. Previously, the call to `ReportUninterestingCall()` was skipped if the `UntypedPerformDefaultAction()` threw an error. Now, `ReportUninterestingCall()` is always called, guaranteeing that the uninteresting call is reported and potentially generates an error.

I had this issue in one of my tests. The test had `EXPECT_ANY_THROW(some_function());`, but on `some_function()`'s error path, it called an unset mock function which itself threw an error. Where normally this error would be reported by bubbling out of the test, it was swallowed by an over-eager `EXPECT_ANY_THROW()` (which admittedly should have been a `EXPECT_THROW()` in retrospect).
